plutus-tx/src/PlutusTx/Trace.hs

```diff
@@ -1,18 +1,57 @@
+{-# LANGUAGE OverloadedStrings #-}
 {-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
 module PlutusTx.Trace (
   trace,
+  errorTag,
   traceError,
+  traceErrorPartial,
   traceIfFalse,
   traceIfTrue
   ) where
 
 import           PlutusTx.Bool
 import           PlutusTx.Builtins as Builtins
 
+{- Note [traceError, traceErrorPartial, and errorTag]
+   To allow QuickCheck to check that validation isn't failing because a partial
```

Comment:
I think one thing that I'm still fundamentally confused about is: it's not clear to me that this distinction is principled. _Why_ are partial functions provided by us bad, but errors thrown by the user okay? Maybe what you're saying is: the only allowable errors should be those that the user thinks indicate "normal" failure modes in their program. But if we look at it that way, I think it makes more sense to ask the user to tag their "normal" errors.

For example, as a user I might use `error` to indicate a condition that's "impossible", i.e. indicates a bug in my code. Such a thing should *not* be considered an okay error.

So what about a system like this: we provide `checkNoUnexpectedErrors`, which takes a `pred :: Text -> Bool` which is provided _by the user_ as a condition on what the last log must be if there's an error.

Then the user can do whatever they want:

```
throwMyError :: BuiltinString -> a -> a
throwMyError s = traceError("Quviq4Eva:" `Builtins.appendString` str)

checkNoUnexpectedErrors (\l -> "Quiviq4Eva" `prefix` l)
```

Or they could compare against a fixed list of error codes (which would double up as a check that your list of error codes was exhaustive!). Indeed, having an exhaustive list of possible error cases might be exactly the sort of thing that you want to insist on for certification!

I think something like this is a) more flexible, b) not *that* much harder to use, c) doesn't mess with any of the prelude code at all.

WDYT?